cidc_api/resources/downloadable_files.py

```diff
@@ -0,0 +1,137 @@
+from ..shared.rest_utils import (
+    lookup,
+    marshal_response,
+    unmarshal_request,
+    use_args_with_pagination,
+)
+
+downloadable_files_bp = Blueprint("downloadable_files", __name__)
+
+downloadable_files_schema = DownloadableFileSchema()
+downloadable_files_list_schema = DownloadableFileListSchema()
+
+
+file_filter_params = {
+    "trial_ids": fields.DelimitedList(fields.Str(), default=[]),
+    "upload_types": fields.DelimitedList(fields.Str(), default=[]),
+    "analysis_friendly": fields.Bool(default=False),
+}
+
+
+@downloadable_files_bp.route("/", methods=["GET"])
+@requires_auth("downloadable_files")
+@use_args_with_pagination(file_filter_params, downloadable_files_schema)
+@marshal_response(downloadable_files_list_schema)
+def list_downloadable_files(args, pagination_args):
+    """List downloadable files that the current user is allowed to view."""
+    user = get_current_user()
+    non_admin_user_id = user.id if not user.is_admin() else None
+
+    filter_ = DownloadableFiles.build_file_filter(
+        **args, non_admin_user_id=non_admin_user_id
+    )
+
+    files = DownloadableFiles.list(filter_=filter_, **pagination_args)
+    count = DownloadableFiles.count(filter_=filter_)
+
+    return {"_items": files, "_meta": {"total": count}}
+
+
+@downloadable_files_bp.route("/<int:downloadable_file>", methods=["GET"])
+@requires_auth("downloadable_files")
+@lookup(DownloadableFiles, "downloadable_file")
+@marshal_response(downloadable_files_schema)
+def get_downloadable_file(downloadable_file: DownloadableFiles) -> DownloadableFiles:
+    """Get a single file by ID if the current user is allowed to view it."""
+    user = get_current_user()
+
+    # Admins can view any file
+    if user.is_admin():
+        return downloadable_file
+
+    # Check that a non-admin has permission to view this file
+    perm = Permissions.find_for_user_trial_type(
+        user.id, downloadable_file.trial_id, downloadable_file.upload_type
+    )
+
+    if not perm:
+        raise NotFound()
+
+    return downloadable_file
+
+
+@downloadable_files_bp.route("/download_url", methods=["GET"])
+@requires_auth("download_url")
+@use_args({"id": fields.Str(required=True)}, location="query")
+def get_download_url(args):
+    """
+    Get a signed GCS download URL for a given file.
+    """
+    # Extract file ID from route
+    file_id = args["id"]
+
+    # Check that file exists
+    file_record = DownloadableFiles.find_by_id(file_id)
+    if not file_record:
+        raise NotFound(f"No file with id {file_id}.")
+
+    user = get_current_user()
+
+    # Ensure user has permission to access this file
+    perms = Permissions.find_for_user(user.id)
+    if user.role != CIDCRole.ADMIN.value:
+        # Check for a permission matching this file's trial and assay
+        if not any(
+            perm.upload_type == file_record.upload_type
+            and perm.trial_id == file_record.trial_id
+            for perm in perms
+        ):
```

Comment:
maybe it would be safer to have `Permissions` method that will only return perms for specific upload type and trial id? Than this (little bit) tricky thing here would become easier to read/understand and not to screw up (not that I'm saying it's screwed). Also getting from the DB only needed to check perms would probably work faster.